src/DistributedWriteAheadLog/Record.cpp

```diff
@@ -13,18 +15,23 @@ namespace DB
 {
 namespace DWAL
 {
-ByteVector Record::write(const Record & record)
+ByteVector Record::write(const Record & record, const bool compressed)
 {
     ByteVector data{static_cast<size_t>((record.block.bytes() + 2) * 1.5)};
     WriteBufferFromVector wb{data};
-    NativeBlockOutputStream output(wb, 0, Block{});
+    auto may_compressed_out = compressed ? CompressedWriteBuffer(wb) : wb;
```

Comment:
There are couple of things.
1. If we encode compression in the `flags`, we shall write the flags without compression and then compression data. The layout will be `[flags uncompressed][compressed or uncompressed binary data]`
2. This is critical path, we shall be really careful about what we do here like memory allocation and we will need enhance the unit test.

Reply:
3. We shall disable Kafka broker compression if we do client side compression
4. There is another way to do this is via kafka message headers instead of flags.

Reply:
A possible rewrite

```
    ByteVector data{static_cast<size_t>((record.block.bytes() + 2) * 1.5)};
    WriteBufferFromVector wb{data};
    
    UInt64 flags = VERSION | (static_cast<UInt8>(record.op_code) << 5ul);
    flags |= ((UInt64) compressed) << 11ul;
    writeIntBinary(flags, wb);
    
    if (unlikely(compressed))
    {
        NativeBlockOutputStream output(CompressedWriteBuffer(wb), 0, Block{});
        output.write(record.block);
        output.flush();
    }
    else
    {
        NativeBlockOutputStream output(wb, 0, Block{});
        output.write(record.block);
        output.flush();
    }
    wb.finalize();
    return data;
```